### PR TITLE
Bins cleanup and enhancements

### PR DESCRIPTION
**Pull request by SimonHeybrock**
- Modify API on Python side as discussed.
  - The odd `groupby['x']` is gone now.
  - Requiring `.bins` in more places.
  - Add `sc.lookup`
- Rename some files buckets - > bins (more to do in the future).
- Some internal cleanup and enhancements for `histogram`.
- Start to modify `sc.bin` (rather the to be renamed `bucketby` on the C++ side) to work with non-edge coords. Note that the corresponding function signatures are not as discussed yet, so that part should be ignored (not how the group assignment works though).

### CODE REVIEW

File: docs/tutorials/neutron-diffraction.ipynb [hunk 1]
```diff
@@ -296,7 +296,13 @@
    "metadata": {},
    "outputs": [],
    "source": [
+<<<<<<< HEAD
```
Comment (Matthew-Andrew) [inline docs/tutorials/neutron-diffraction.ipynb:299]: Looks like some unresolved merge conflicts?

File: python/src/scipp/_bins.py [hunk 2]
```diff
@@ -52,6 +26,24 @@ def _data(self):
...
+        _cpp.buckets.scale(copy, _cpp.reciprocal(lut.func), lut.dim)
+        return copy
+
+    def __imul__(self, lut: lookup):
+        _cpp.buckets.scale(self._obj, lut.func, lut.dim)
+        return self
+
+    def __itruediv__(self, lut: lookup):
+        _cpp.buckets.scale(self._obj, _cpp.reciprocal(lut.func), lut.dim)
+        return self
```
Comment (OwenArnold) [inline python/src/scipp/_bins.py:45]: 👍

File: docs/tutorials/neutron-diffraction.ipynb [hunk 3]
```diff
@@ -401,13 +401,14 @@
...
-    "# Therefore we first need to concatenate bins along d-spacing before concatenateing spectra.\n",
+    "# Therefore we first need to concatenate bins along d-spacing before concatenating spectra.\n",
     "tmp = dspacing_sample.bins.concatenate('d-spacing')\n",
     "focussed_sample = sc.groupby(tmp, 'theta', bins=theta).bins.concatenate('spectrum')\n",
     "tmp = dspacing_vanadium.bins.concatenate('d-spacing')\n",
     "focussed_vanadium = sc.groupby(tmp, 'theta', bins=theta).bins.concatenate('spectrum')\n",
-    "focussed_sample.bins.groupby['d-spacing'] /= sc.histogram(focussed_vanadium, dspacing_bins)\n",
-    "normalized = sc.histogram(focussed_sample, dspacing_bins)"
+    "norm = sc.histogram(focussed_vanadium.bins, dspacing_bins)\n",
+    "focussed_sample.bins /= sc.lookup(func=norm, dim='d-spacing')\n",
```
Comment (OwenArnold) [inline docs/tutorials/neutron-diffraction.ipynb:410]: If monitors were run in event mode would it be necessary to histogram them before this lookup approach was possible? While histogram monitors is usually what is required, there are legitimate cases where monitors are recorded in event mode.

**Review by OwenArnold — APPROVED** [review]

File: docs/tutorials/neutron-diffraction.ipynb (hunk 3 above)
Comment (SimonHeybrock) [reply docs/tutorials/neutron-diffraction.ipynb:410]: Yes, see https://scipp.github.io/event-data/overview.html#Arithmetic-operations. Only addition and subtraction are supported if both operands are event data.

File: docs/event-data/filtering.ipynb [hunk 4]
```diff
@@ -258,7 +258,7 @@
    "source": [
     "tof_bins = sc.Variable(dims=['tof'], unit=sc.units.us, values=np.linspace(0,tof_max,num=100))\n",
     "temp_bins = sc.Variable(dims=['temperature'], unit=sc.units.K, values=np.linspace(100.0, 130.0, num=6))\n",
-    "binned_events = sc.bin(events, [temp_bins, tof_bins])\n",
+    "binned_events = sc.bin(events.bins, [temp_bins, tof_bins])\n",
```
[the remark this answers is not in the document]
Comment (SimonHeybrock) [reply docs/event-data/filtering.ipynb:261]: Will, as soon as our confidence about whether this is the correct API not has increased.

File: docs/event-data/overview.ipynb [hunk 5]
```diff
@@ -289,9 +311,8 @@
    },
    "outputs": [],
    "source": [
-    "time_bins = sc.Variable(dims=['time'], unit=sc.units.us, values=[0.0, 3.0, 6.0])\n",
-    "a.bins.groupby['time'] /= sc.histogram(a, time_bins)\n",
-    "plot(sc.histogram(a, time_bins))"
+    "a.bins /= sc.lookup(func=hist, dim='time')\n",
+    "plot(sc.histogram(a.bins, time_bins))"
```
[the remark this answers is not in the document]
Comment (SimonHeybrock) [reply docs/event-data/overview.ipynb:315]: Yes, it should.

File: docs/scipp-neutron/groupby.ipynb [hunk 6]
```diff
@@ -138,7 +138,7 @@
    "metadata": {},
    "outputs": [],
    "source": [
-    "sc.plot.plot(sc.histogram(events, bins))"
+    "sc.plot.plot(sc.histogram(events.bins, bins))"
```
[the remark this answers is not in the document]
Comment (SimonHeybrock) [reply docs/scipp-neutron/groupby.ipynb:141]: This PR was opened before plotting supported binning, but as you say maybe staying explict is good?

**Review by nvaytet — APPROVED** [review]

**Pull request merged**